### PR TITLE
Pick API versions when sending requests to kafka

### LINKED ISSUE #158: when kafka version is 0.10.0.0 and brod send consume protocol in version 0.9 or 0.8, may this cause performance impact?
description is in 
http://kafka.apache.org/documentation/#upgrade_10_performance_impact

```
The message format in 0.10.0 includes a new timestamp field and uses relative offsets for compressed messages. The on disk message format can be configured through log.message.format.version in the server.properties file. The default on-disk message format is 0.10.0. If a consumer client is on a version before 0.10.0.0, it only understands message formats before 0.10.0. In this case, the broker is able to convert messages from the 0.10.0 format to an earlier format before sending the response to the consumer on an older version. However, the broker can't use zero-copy transfer in this case. Reports from the Kafka community on the performance impact have shown CPU utilization going from 20% before to 100% after an upgrade, which forced an immediate upgrade of all clients to bring performance back to normal. To avoid such message conversion before consumers are upgraded to 0.10.0.0, one can set log.message.format.version to 0.8.2 or 0.9.0 when upgrading the broker to 0.10.0.0. This way, the broker can still use zero-copy transfer to send the data to the old consumers. Once consumers are upgraded, one can change the message format to 0.10.0 on the broker and enjoy the new message format that includes new timestamp and improved compression. The conversion is supported to ensure compatibility and can be useful to support a few apps that have not updated to newer clients yet, but is impractical to support all consumer traffic on even an overprovisioned cluster. Therefore, it is critical to avoid the message conversion as much as possible when brokers have been upgraded but the majority of clients have not.
```

### LINKED ISSUE #205: Support for Message Timestamp
Is the Kafka [message timestamp field](https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol) something Brod could expose in the [message record](https://github.com/klarna/brod/blob/master/include/brod.hrl#L33)?

From the docs, it's hard to understand if this is actually part of the protocol at this point (it refers to 0.10.0) or something that's not there yet.

Thanks!

### PR DESCRIPTION
**Pull request by zmstone**
Fixes #158 
Fixes #205

### CODE REVIEW

**Review by id — COMMENTED** [review]
Don't know what exactly is wrong, but on this version metadata request with `topics = []` returns no info on topics

**Comment by zmstone** [comment]
good catch, the 'all' topics indicator has been changed from 'empty array' to 'null' since version 1.

**Comment by zmstone** [comment]
need to release a 1.1 for `kafka_protocol` first.

File: src/brod_kafka_apis.erl [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+  {Min, _Max} = supported_versions(API),
+  Min.
+
+%% @doc Pick API version for the given API.
+-spec pick_version(pid(), api()) -> vsn().
+pick_version(SockPid, API) ->
+  do_pick_version(SockPid, API, supported_versions(API)).
+
+init([]) ->
+  ets:new(?ETS, [named_table, protected, duplicate_bag]),
```
[the remark this answers is not in the document]
Comment (zmstone) [reply src/brod_kafka_apis.erl]: the key to this table is pid(), so it should never duplicate.

File: src/brod_consumer.erl [hunk 2]
```diff
@@ -132,6 +133,10 @@ start_link(ClientPid, Topic, Partition, Config) ->
 %%                          'begin_offset' option.
 %%     reset_to_earliest: consume from the earliest offset.
 %%     reset_to_latest: consume from the last available offset.
+%%  size_stat_window: (optional, default = 5)
+%%     The moving-average window size when shrinking `max_bytes` after it has
+%%     been expanded to fetch a large message. Use 0 to immediately shrink
+%%     back to original `max_bytes` from config.
```
Comment (id) [inline src/brod_consumer.erl]: use regular single quotes, not backticks

File: src/brod_kafka_apis.erl (hunk 1 above)
[the remark this answers is not in the document]
Comment (id) [reply src/brod_kafka_apis.erl]: then regular `set` should do just fine, or?

[the remark this answers is not in the document]
Comment (zmstone) [reply src/brod_kafka_apis.erl]: yes, but duplicate_bag has the best performance

[the remark this answers is not in the document]
Comment (id) [reply src/brod_kafka_apis.erl]: For insert yes, but we need to optimize for lookups, or? Lookups are more efficient in `set`, take this one for example: https://www.technomenace.com/2016/01/erlang-ets-types-readwrite-performance-tests/

[the remark this answers is not in the document]
Comment (zmstone) [reply src/brod_kafka_apis.erl]: d'oh
I googled to the same page, but took traversing performance as lookup performance.

**Review by id — APPROVED** [review]

**Pull request merged**